mage_ai/io/cratedb.py

```diff
@@ -0,0 +1,389 @@
+                    default=encode_complex,
+                    ignore_nan=True,
+                ))
+            return val
+
+        df_ = df.copy()
+        columns = df_.columns
+
+        for col in columns:
+            df_col_dropna = df_[col].dropna()
+            if df_col_dropna.count() == 0:
+                continue
+            if dtypes[col] == PandasTypes.OBJECT \
+                    or (df_[col].dtype == PandasTypes.OBJECT and not
+                        isinstance(df_col_dropna.iloc[0], str)):
+                df_[col] = df_[col].apply(lambda x: serialize_obj(x))
+        df_.replace({np.NaN: None}, inplace=True)
+
+        insert_columns = ', '.join([f'"{col}"'for col in columns])
+
+        if use_insert_command:
+            # Use INSERT command
+            values_placeholder = ', '.join(["%s" for i in range(len(columns))])
+            values = []
+            for _, row in df_.iterrows():
+                values.append(tuple(row))
+            commands = [
+                f'INSERT INTO {full_table_name} ({insert_columns})',
+                f'VALUES ({values_placeholder})',
+            ]
+
+            unique_constraints = \
+                [f'"{self._clean_column_name(col, allow_reserved_words=allow_reserved_words)}"'
+                 for col in unique_constraints]
+            columns_cleaned = \
+                [f'"{self._clean_column_name(col, allow_reserved_words=allow_reserved_words)}"'
+                 for col in columns]
+
+            commands.append(f"ON CONFLICT ({', '.join(unique_constraints)})")
+            if UNIQUE_CONFLICT_METHOD_UPDATE == unique_conflict_method:
+                update_command = [f'{col} = EXCLUDED.{col}' for col in columns_cleaned]
+                commands.append(
+                    f"DO UPDATE SET {', '.join(update_command)}",
+                )
+            else:
+                commands.append('DO NOTHING')
+            cursor.executemany('\n'.join(commands), values)
+        else:
+            # Use COPY command
+            df_.to_csv(
+                buffer,
+                header=False,
+                index=False,
+                na_rep='',
+            )
+            buffer.seek(0)
+            cursor.copy_expert(f"""
+COPY {full_table_name} ({insert_columns}) FROM STDIN (
+    FORMAT csv
+    , DELIMITER \',\'
+    , NULL \'\'
+    , FORCE_NULL({insert_columns})
+);
+        """, buffer)
+
+    def execute(self, query_string: str, **query_vars) -> None:
+        """
+        Sends query to the connected database.
+
+        Args:
+            query_string (str): SQL query string to apply on the connected database.
+            query_vars: Variable values to fill in when using format strings in query.
+        """
+        with self.printer.print_msg(f'Executing query \'{query_string}\''):
+            query_string = self._clean_query(query_string)
+            with self.conn.cursor() as cur:
+                cur.execute(query_string, query_vars)
```

Comment:
The `bulk_parameters` argument may want to be used here when applicable, depending on context.

Reply:
Forget about this comment. If it's psycopg, then there isn't such an argument.